enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,209 @@
+
+OLM should not delete CRs of the type [required APIService][required-apiservices] as those are likely created by the operator, and whose cleanup should be handled by the operator's finalizer.
+
+#### Finalizers on CSVs
+
+OLM can initiate the process of operand clean up when the operator is uninstalled, which is when the CSV is deleted. However, deleting a CSV causes the operator deployment (owned by the CSV) to be garbage collected.
+
+So in order to keep the operator around to process CR deletion events, OLM can add a finalizer to the CSV to prevent the CSV and the operator from being removed. Once OLM has completed operand cleanup and can verify the removal of all CRs, the CSV finalizer can be cleared so that the operator can be safely removed.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+metadata:
+  ...
+  finalizers:
+  - olm/operand-cleanup
+```
+
+In the event of operand cleanup being stalled due to some operator error, OLM will not clear the CSV finalizer so that the operator stays alive indefinitely. This way the operator can be observed to figure out the required action to unblock cleanup.
+
+#### Operand cleanup API
+
+The CSV spec can have an optional field `spec.cleanup` which lets a user configure the cleanup behaviour before the CSV is deleted.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+spec:
+  ...
+  cleanup:
+    enabled: false
+```
+
+With `spec.cleanup.enabled: true` OLM will first add the `olm/operand-cleanup` finalizer on the CSV. Now when the CSV is deleted by the user, OLM will delete all CRs for the operator's owned APIs, in the target namespaces of that operator. Once the CRs are successfully removed (after the operator clears its finalizers), OLM will clear the finalizer from the CSV to let the CSV and operator be uninstalled.
+
+With `spec.cleanup.enabled: false` OLM will remove the `olm/operand-cleanup` finalizer if it is already present on the CSV. This allows users to opt-out of cleanup and not have OLM delete any CRs.
+
+This is also the default behavior if `spec.cleanup` is unspecified. This ensures operator's that haven't opted-in are not subjected to the finalizer and have their CSV lifecycle remain unchanged.
+
+While toggling `spec.cleanup.enabled` does not result in a phase change, OLM could generate events to provide a history of when the cleanup behavior was toggled, and also events for when OLM begins and ends the cleanup process.
+
+#### Operand cleanup status
+
+To provide visibility on the status of operand cleanup, OLM can list out the CRs that are still pending deletion in the CSV status e.g `status.cleanup.pendingDeletion`.
```

Comment:
are we certain the k8s API lets us update the status of a resource that has a deletionTimestamp set?

Reply:
I can't find explicit docs for this but from a quick test it seems like you can update the status (or any other field) even after DeletionTimestamp is set and the resource remains visible in the foreground due to a finalizer.